src/ast_util.py

```diff
@@ -166,6 +166,126 @@ def generic_visit(self, node: ca.Node) -> None:
+    new_name = name
+    counter = 1
+    while new_name in used_names:
+        counter += 1
+        new_name = f"{new_name}{counter}"
+    return new_name
+
+
+def ast_ext_struct_idx(ast: ca.FileAST, struct: ca.Struct) -> int:
+    """Returns the index of the node in `ast.ext` that contains the definition
+    of `struct`.
+    """
+    found = False
+    class StructVisitor(ca.NodeVisitor):
+        def visit_Struct(self, node: ca.Struct) -> None:
+            nonlocal found
+            if node is struct:
+                found = True
+            self.generic_visit(node)
+
+    for i, node in enumerate(ast.ext):
+        StructVisitor().visit(node)
+        if found:
+            return i
+    return -1
+
+
+def copy_struct(ast: ca.FileAST, typemap: TypeMap, struct: ca.Struct) -> ca.Struct:
+    """Adds a new struct with a definition matching `struct` to the AST.
+    Returns this new struct.
+    """
+    new_struct = copy.deepcopy(struct)
+    base_name = struct.name if struct.name else struct.decls[0].name
+    new_struct.name = get_noncolliding_struct_name(typemap, base_name)
+
+    # Insert the new struct right above the base struct. This preserves any
+    # struct field dependencies.
+    insert_idx = ast_ext_struct_idx(ast, struct)
+    decl = ca.Decl(name=None, quals=[], align=[], storage=[], funcspec=[], type=new_struct, init=None, bitsize=None)
+    ast.ext = ast.ext[:insert_idx] + [decl] + ast.ext[insert_idx:]
+    return new_struct
+
+
+def struct_set_field_type(parent: ca.Struct, field_idx: int, new_type: ca.Struct) -> None:
+    struct_type = copy.deepcopy(new_type)
+    struct_type.decls = None
+    old_type_decl = parent.decls[field_idx].type
+    assert isinstance(old_type_decl, ca.TypeDecl), "can only modify the type of a TypeDecl field"
+    old_type_decl.type = struct_type
+
+
+def struct_substruct_idx(parent: ca.Struct, substruct: ca.Struct, typemap: TypeMap) -> int:
+    """Returns the index of the field in `parent` of type `substruct`.
+    """
+    for i, decl in enumerate(parent.decls):
+        if isinstance(decl.type, ca.TypeDecl):
+            decl_struct = resolve_struct_def(decl.type, typemap)
+            if decl_struct is substruct:
+                return i
+    return -1
+
+
+def struct_rename_field(field: ca.Decl, name: str) -> None:
+    field.name = name
+    class TypeDeclVisitor(ca.NodeVisitor):
+        def visit_TypeDecl(self, node: ca.TypeDecl) -> None:
+            node.declname = name
+
+    TypeDeclVisitor().visit(field)
+
+
+def struct_remove_field(struct: ca.Struct, field_idx: int) -> None:
+    struct.decls = struct.decls[:field_idx] + struct.decls[field_idx+1:]
+
+
+def count_chained_structrefs(node: ca.StructRef) -> int:
+    """Count the number of chained StructRef's that are required to reach the
+    field in `node`.
+
+    Examples:
+      - foo            --> 0
+      - bar.foo        --> 1
+      - baz.bar.foo    --> 2
+      - bar->foo       --> 0
```

Comment:
I'm getting 1 for this trying to trace the code: the code looks at `node.type` instead of `node.name.type` so it does traverse all the way to the -> node before bailing, adding +1 along the way. Guessing that's unintentional?

Reply:
I think this actually should be 0, following the new implementation. This expression maps to
```
StructRef: ->
  ID: bar
  ID: foo
```

Here, `expr.type` will be "->", so we'll never enter the loop.